### ISSUE TITLE
Create an architecture diagram

### ISSUE THREAD

**Issue by BjoernPetersen** [tim-hat-die-hand-an-der-maus/architecture#8]

**Comment by torbencarstens** [comment]
Look at Mr. Fancypants over here wanting an architecure diagram

I agree though, it'll become too complicated at some point. some more microservices in.

**Comment by BjoernPetersen** [comment]
I'm confused already! (at times)

**Comment by torbencarstens** [comment]
Getting old is tough.

I think we should have more than one architecure diagram:
- the group of stakeholders is probably more interested in a rough overview
- a new frontend developer is mostly interested in the API connections
- [...]

I've created a rough overview, I left some parts as a blackbox to provide a better overview.

We can expand on this after we've discussed the scope of the different architecture diagrams:

![image](https://user-images.githubusercontent.com/7738181/157731677-867b5a15-874c-45ca-89c4-461c09abee4b.png)

**Comment by BjoernPetersen** [comment]
Nice! I think that covers the most important parts.

**Comment by torbencarstens** [comment]
Since I've recently ran into this problem I'll give a quick rundown of the services, I'll leave out the bus-frontend since it's a self-contained site. This is the current state of affairs, planned features can added later.

### api

Has all the information (connects to postgres)

### preliminary-frontend

read-only, shows all unwatched movies

Connects to: api

### telegram-bot

Current frontend which provides basic functionality for the user (add, delete, ...)

Connects to: api

### plex-resolver

Provides an api endpoint which returns all movies (plex library type: Movies) from the configured plex servers as json

Currently unused 

### plex-requester

Other services can send a request for a movie to be added to the download queue

Backend: to be decided

In development

### streaming-provider-resolver

Checks which streaming service (info by werstreamt.es)has which movue and returns this information from an api endpoint

Currently unused

### ImdB-resolver

Provides an api which returns information from ImdB about the title requested

### metadata-updater

Checks for outdated Metadata (e.g. 404 for thumbnails) and tells the api to request new Metadata from the ImdB-resolver

Connects to: ImdB-resolver, api

### rotating-docx-file-handler

Best logging impl out there

**Comment by torbencarstens** [comment]
I'm gonna take another stab at actually creating a diagram, I'll try out a few different libraries/services for doing this.

First up, Ilograph: https://app.ilograph.com/@torben/tim/Dependency

**Comment by torbencarstens** [comment]
![d2](https://github.com/tim-hat-die-hand-an-der-maus/architecture/assets/7738181/4f7908fb-6e92-408d-a559-e8d93e5bc983)

example from d2:

<details>
<summary>d2lang code</summary>
<pre>users -> preliminary-frontend
users -> telegram-bot

preliminary-frontend -> api: Read
telegram-bot -> api: Read/Write
telegram-bot -> imdb-resolver: Read
metadata-updater -> imdb: Read
streamingprovider-resolver -> werstreamtes: Read
metadata-updater -> imdb-resolver: Read
metadata-updater -> imdb: Read
metadata-updater -> api: Read/Write
plex-resolver -> plex-request: Write
streamingprovider-resolver -> plex-request: Read
telegram-bot -> plex-request: Write
imdb-resolver -> imdb: Read
api -> postgres: Read/Write</pre>
</details>

**Comment by BjoernPetersen** [comment]
Doesn't the API make requests to the imdb resolver?

**Comment by BjoernPetersen** [comment]
I'm impressed that no other service has snuck in direct access to the database yet. It's definitely a beautiful mess that we can only make worse.

**Comment by torbencarstens** [comment]
> Doesn't the API make requests to the imdb resolver?
> 
> Edit: yes it does. The metadata-updater just tells the API to refresh certain fields of a movie. I think the idea was somehow that the API needs to talk to the imdb-resolver anyway for new movies, so we didn't want to duplicate the imdb-resolver client code in another service.

huh, didn't even know this was the case. I've added this as an additional dependency.

While ilographs [pricing model](https://www.ilograph.com/pricing.html) is absolutely insane (especially for our use-case), it still has a very nice set of features; specifically being able to view different perspectives depending on service/parent resource/...

Which makes [this part](https://github.com/tim-hat-die-hand-an-der-maus/architecture/issues/8#issuecomment-1064372335):

> I think we should have more than one architecure diagram:
>
> - the group of stakeholders is probably more interested in a rough overview
> - a new frontend developer is mostly interested in the API connections
> - [...]

fairly trivial.

The export API (exporting standalone HTML) [is even more ridiculous](https://www.ilograph.com/export-api/)

Downside: We're dependent on their service continuation and can only update diagrams manually.

That being said, after the initial completion we'll proabably update this at most once a month (on average) and migrating to another language (such as d2) for the individual perspectives isn't that hard.

I've started this here: https://app.ilograph.com/@torben/kubernetes/

I'll play around a bit with the different concepts so this graph may change/misbehave at any time.

> I'm impressed that no other service has snuck in direct access to the database yet. It's definitely a beautiful mess that we can only make worse.

Yeah, I'm amazed too. I think we only violate this (in all of our projects) in `woog-life/graphing` since the API doesn't expose every entry.

**Comment by torbencarstens** [comment]
> > huh, didn't even know this was the case. I've added this as an additional dependency.
> 
> You actually knew once: [#8 (comment)](https://github.com/tim-hat-die-hand-an-der-maus/architecture/issues/8#issuecomment-1091945965)

Do you really expect me to r*ad?

I _may_ [share your opinion though](https://github.com/tim-hat-die-hand-an-der-maus/architecture/issues/8#issuecomment-1063794397).
 
> > ### metadata-updater
> > Checks for outdated Metadata (e.g. 404 for thumbnails) and tells the api to request new Metadata from the ImdB-resolver
> 
> But I wouldn't have remembered that either, [we're all getting old](https://github.com/tim-hat-die-hand-an-der-maus/architecture/issues/8#issuecomment-1063794397).

I had to look at the code to check how it works since the last (code related) commit I've made was March 2022.

But there aren't any missing thumbnails (which was the usual problem) so I guess this works.

> > Downside: We're dependent on their service continuation and can only update diagrams manually.
> 
> I'm really hesitant to buy into a lock-in like that, but in the end, you're the one spending all that time on it now, so it's mostly your loss if we need to migrate at some point 😬.

Yeah, me too. I'll probably write a small tool which converts the ilograph structure to a basic d2-lang structure.

basically

```
from: x
to: y
```

simply converts to

```
x -> y
```

for each perspective.
